### PR TITLE
Fix/safe indexing bound check

### LINKED ISSUE #34: Out of bounds access due to incorrect pre-check in register_enclave causes panic
# Summary

The extrinsic call `register_enclave` in `pallet-teerex/src/lib.rs` leads to an out of bound memory access in `safe_indexing_one` function in `pallet-teerex/ias-verify/src/lib.rs`. As this can be triggered via an extrinsic call, an attacker could perform a DoS attack against the whole chain by exploiting this vulnerability. 

# Issue

In `pallet-teerex/src/lib.rs`, the extrinsic `register_enclave` calls into `verify_ias_report` which in turn calls into `safe_indexing_one` function in `pallet-teerex/ias-verify/src/lib.rs` which fails to properly perform bounds checking and causes a panic when `data.len() == idx`.

```rust
fn safe_indexing_one(data: &[u8], idx: usize) -> Result<u8, &'static str> {
    if data.len() < idx {
        return Err("Index would be out of bounds");
    }
    Ok(data[idx])
}
```

The following extrinsic call can trigger this vulnerability:

```rust
Call::SubstrateeRegistry(register_enclave([0, 0, 0, 8, 255, 6, 8, 42, 134, 72, 206, 61, 3, 1, 7, 48], [0, 0, 0, 5, 0, 1, 15, 55, 8, 1, 0, 0, 0, 1, 0, 53, 51, 56, 49, 49, 52, 0, 0, 0, 0, 0, 0, 0, 0, 0]))
```

# Risk

Since this can be triggered by a call to an extrinsic, an attacker could crash a validator by submitting a specially crafted extrinsic and cause them miss their block authoring slot. This could lead **to a denial-of-service of the whole chain**.

# Mitigation

Fix the bounds checking code:

```rust
fn safe_indexing_one(data: &[u8], idx: usize) -> Result<u8, &'static str> {
    if data.len() <= idx {
        return Err("Index would be out of bounds");
    }
    Ok(data[idx])
}
```

### PR DESCRIPTION
**Pull request by clangenb**
* Closes #34

Todo:

- [x]  Merge #36 
- [x] rebase on master
- [x] retarget this PR to master

### CODE REVIEW

File: ias-verify/src/utils.rs [hunk 1]
```diff
@@ -4,14 +4,14 @@ pub fn safe_indexing(data: &[u8], start: usize, end: usize) -> Result<&[u8], &'s
...
+        return Err("Index out of bounds");
     }
     Ok(&data[start..end])
 }
 
 pub fn safe_indexing_one(data: &[u8], idx: usize) -> Result<u8, &'static str> {
-    if data.len() < idx {
-        return Err("Index would be out of bounds");
+    if data.len() <= idx {
+        return Err("Index out of bounds");
```
Comment (mullefel) [inline ias-verify/src/utils.rs:14]: Good catch! probably by writing unit tests? 😄 

Does this functionality not already exist somewhere in the standard library? Seems surprising to me, that we have to write code like this - is there not a `.get()` method on `slice` that returns an `Option`?

File: ias-verify/src/utils.rs [hunk 2]
```diff
@@ -4,14 +4,14 @@ pub fn safe_indexing(data: &[u8], start: usize, end: usize) -> Result<&[u8], &'s
         return Err("Illegal indexing");
     }
     if data.len() < end {
-        return Err("Index would be out of bounds");
+        return Err("Index out of bounds");
     }
     Ok(&data[start..end])
```
Comment (mullefel) [inline ias-verify/src/utils.rs:9]: Also here, can't we just use `slice.get()` for exactly the same functionality?

File: ias-verify/src/utils.rs (hunk 1 above)
Comment (clangenb) [reply ias-verify/src/utils.rs:14]: Would love to take the credits, but this was found in the security review of polkadex.

Oh interesting, I am 99% sure that this did not exist 2 years ago. Will incorporate that. Thanks for the hint!

**Review by brenzi — APPROVED** [review]

File: ias-verify/src/utils.rs (hunk 1 above)
Comment (brenzi) [reply ias-verify/src/utils.rs:14]: You have my review, please merge at your discretion

Comment (clangenb) [reply ias-verify/src/utils.rs:14]: Haha, I forgot. Your link is part of the standard-lib, which is why we can't use it. :D

**Comment by clangenb** [comment]
I have seen that `sp-std` supports slices including the `get` method. I will not add this into this pr. Tracking issue here: #40

**Pull request merged**